classes/property_list.php

```diff
@@ -0,0 +1,295 @@
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Wrapper for CFPropertyList to handle low level iteration.
+ *
+ * @package    quizaccess_seb
+ * @author     Andrew Madden <andrewmadden@catalyst-au.net>
+ * @copyright  2019 Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace quizaccess_seb;
+
+use CFPropertyList\CFArray;
+use CFPropertyList\CFData;
+use CFPropertyList\CFDate;
+use CFPropertyList\CFDictionary;
+use CFPropertyList\CFPropertyList;
+use CFPropertyList\CFString;
+use CFPropertyList\CFType;
+
+defined('MOODLE_INTERNAL') || die();
+
+require_once(__DIR__ . '/../vendor/autoload.php');
+
+class property_list {
+
+    /** @var CFPropertyList $cfpropertylist */
+    private $cfpropertylist;
+
+    /**
+     * property_list constructor.
+     *
+     * @param string $xml A Plist XML string.
+     *
+     * @throws \CFPropertyList\IOException
+     * @throws \CFPropertyList\PListException
+     * @throws \DOMException
+     */
+    public function __construct(string $xml) {
+        $this->cfpropertylist = new CFPropertyList();
+        $this->cfpropertylist->parse($xml, CFPropertyList::FORMAT_XML);
+    }
+
+    /**
+     * Create an empty property list.
+     *
+     * @return property_list Return a property list with a empty root dictionary set up.
+     *
+     * @throws \CFPropertyList\IOException
+     * @throws \CFPropertyList\PListException
+     * @throws \DOMException
+     */
+    public static function create() {
+        $cfpropertylist = new CFPropertyList();
+        // Add main dict.
+        $cfpropertylist->add(new CFDictionary([]));
+        return new self($cfpropertylist->toXML());
+    }
+
+    /**
+     * Add a new element to the root dictionary element.
+     *
+     * @param string $key Key to assign to new element.
+     * @param CFType $element The new element. May be a collection such as an array.
+     */
+    public function add_element_to_root(string $key, CFType $element) {
+        // Get the PList's root dictionary and add new element.
+        $this->cfpropertylist->getValue()->add($key, $element);
+    }
+
+    /**
+     * Get value of element identified by key.
+     *
+     * @param string $key Key of element.
+     * @return mixed Value of element found, or null if none found.
+     */
+    public function get_element_value(string $key) {
+        $result = null;
+        $this->plist_map( function($elvalue, $elkey, $parent) use ($key, &$result) {
+            // Convert date to iso 8601 if date object.
+            if ($key === $elkey) {
+                $result = $elvalue->getValue();
+            }
+        }, $this->cfpropertylist->getValue());
+        // Turn PList dicts and arrays into PHP array for export.
+        if (is_array($result)) {
+            $result = new CFDictionary($result); // Convert back to CFDictionary so serialization is recursive.
```

Comment:
Maybe abstract this initialisation of CFDictionary ?

Reply:
Created new method - `array_serialize_cftypes`